File: solutions/haskell/protein-translation/1/src/ProteinTranslation.hs
module ProteinTranslation (proteins) where

proteins :: String -> Maybe [String]
proteins rna =
    case (rna, lookup codon proteinTable) of
        ("", _) -> Just []
        (_, Just "STOP") -> Just []
        (_, Just protein) -> (protein :) <$> proteins rest
        _ -> Nothing
  where
    (codon, rest) = splitAt 3 rna

proteinTable :: [(String, String)]
proteinTable =
    [ (codon, protein)
    | (codons, protein) <-
        [ (["AUG"], "Methionine")
        , (["UUU", "UUC"], "Phenylalanine")
        , (["UUA", "UUG"], "Leucine")
        , (["UCU", "UCC", "UCA", "UCG"], "Serine")
        , (["UAU", "UAC"], "Tyrosine")
        , (["UGU", "UGC"], "Cysteine")
        , (["UGG"], "Tryptophan")
        , (["UAA", "UAG", "UGA"], "STOP")
        ]
    , codon <- codons
    ]

File: solutions/cpp/protein-translation/1/protein_translation.cpp
#include "protein_translation.h"

#include <map>
#include <optional>
#include <set>

namespace protein_translation {

using namespace std;

optional<string> lookup_protein(string codon) {
    const map<string, set<string>> name_to_codon{
        {"STOP", {"UAA", "UAG", "UGA"}},
        {"Cysteine", {"UGU", "UGC"}},
        {"Leucine", {"UUA", "UUG"}},
        {"Methionine", {"AUG"}},
        {"Phenylalanine", {"UUU", "UUC"}},
        {"Serine", {"UCU", "UCC", "UCA", "UCG"}},
        {"Tryptophan", {"UGG"}},
        {"Tyrosine", {"UAU", "UAC"}},
    };
    static map<string, string> codon_to_name;
    if (codon_to_name.empty()) {
        for (const auto& [name, codons] : name_to_codon) {
            for (const auto& codon : codons) {
                codon_to_name.emplace(codon, name);
            }
        }
    }
    const auto it = codon_to_name.find(codon);
    return it != codon_to_name.end() ? optional{it->second} : nullopt;
}

vector<string> proteins(string rna) {
    vector<string> proteins;
    for (size_t i = 0; i < rna.size() / 3; i++) {
        const auto codon = rna.substr(3 * i, 3);
        const auto mprotein = lookup_protein(codon);
        if (not mprotein) {
            return {};
        }
        if (*mprotein == "STOP") {
            break;
        }
        proteins.push_back(*mprotein);
    }
    return proteins;
}

}  // namespace protein_translation
